test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.cc

```diff
@@ -0,0 +1,303 @@
+      if (fh->immediate_resp_sent_) {
+        ENVOY_LOG_MISC(trace, "Proxy closed connection, returning early");
+        fh->immediate_resp_lock_.unlock();
+        return response;
+      }
+      uint32_t data_size = fdp->ConsumeIntegralInRange<uint32_t>(0, ext_proc_fuzz_max_data_size);
+      ENVOY_LOG_MISC(trace, "Sending chunk of {} bytes", data_size);
+      codec_client_->sendData(encoder, data_size, false);
+      fh->immediate_resp_lock_.unlock();
+    }
+
+    // See comment above
+    fh->immediate_resp_lock_.lock();
+    if (!fh->immediate_resp_sent_) {
+      ENVOY_LOG_MISC(trace, "Sending empty chunk to close stream");
+      Buffer::OwnedImpl empty_chunk;
+      codec_client_->sendData(encoder, empty_chunk, true);
+    }
+    fh->immediate_resp_lock_.unlock();
+    return response;
+  }
+
+  IntegrationStreamDecoderPtr randomDownstreamRequest(FuzzedDataProvider* fdp,
+                                                      ExtProcFuzzHelper* fh) {
+    // From the external processor's view each of these requests
+    // are handled the same way. They only differ in what the server should
+    // send back to the client.
+    // TODO(ikepolinsky): add random flag for sending trailers with a request
+    //   using HttpIntegration::sendTrailers()
+    switch (fdp->ConsumeEnum<HttpMethod>()) {
+    case HttpMethod::GET:
+      ENVOY_LOG_MISC(trace, "Sending GET request");
+      return sendDownstreamRequest(absl::nullopt);
+    case HttpMethod::POST:
+      if (fdp->ConsumeBool()) {
+        ENVOY_LOG_MISC(trace, "Sending POST request with body");
+        uint32_t data_size = fdp->ConsumeIntegralInRange<uint32_t>(0, ext_proc_fuzz_max_data_size);
+        std::string data = std::string(data_size, 'a');
+        return sendDownstreamRequestWithBody(data, absl::nullopt);
+      } else {
+        ENVOY_LOG_MISC(trace, "Sending POST request with chunked body");
+        return sendDownstreamRequestWithChunks(fdp, fh, absl::nullopt);
+      }
+    default:
+      ENVOY_LOG_MISC(error, "Unhandled HttpMethod");
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  envoy::extensions::filters::http::ext_proc::v3alpha::ExternalProcessor proto_config_{};
+  TestProcessor test_processor_;
+  Network::Address::IpVersion ip_version_;
+  Grpc::ClientType client_type_;
+};
+
+DEFINE_FUZZER(const uint8_t* buf, size_t len) {
+  // Split the buffer into two buffers with at least 1 byte
+  if (len < 2) {
+    return;
+  }
+
+  // External Process and downstream are on different threads so they should
+  // have separate data providers
+  size_t downstream_buf_len = len / 2;
+  size_t ext_proc_buf_len = len - downstream_buf_len;
+
+  // downstream buf starts at 0, ext_prob buf starts at buf[downstream_buf_len]
+  FuzzedDataProvider downstream_provider(buf, downstream_buf_len);
+  FuzzedDataProvider ext_proc_provider(&buf[downstream_buf_len], ext_proc_buf_len);
+
+  ExtProcIntegrationFuzz fuzzer(Network::Address::IpVersion::v4, Grpc::ClientType::GoogleGrpc);
```

Comment:
@htuch to build off the changes to support IpV6 addressed above, I have a question about how I should handle initializing the ExtProcIntegrationFuzz class. 

Is there a fuzz equivalent of INSTANTIATE_TEST_CASE that is seen in the other integration tests? If not should I set the IpVersion and GrpcVersion randomly with the fuzz provider?

Reply:
In my latest change I initialize the fuzzer using the TestEnvironment functions, allowing environment variables to control these parameters.